File: Plaintext_Attack.java
// Bismillahir Rahmanir Rahim
// Plain-text Attack on RSA

import java.math.BigInteger;

public class Plaintext_Attack {
    static BigInteger power(BigInteger b, BigInteger p, BigInteger mod) {
        b = b.mod(mod);
        if (p.equals(BigInteger.ZERO))
            return BigInteger.ONE;
        BigInteger j = power(b, p.divide(BigInteger.valueOf(2)), mod);
        j = (j.multiply(j)).mod(mod);
        if (p.mod(BigInteger.valueOf(2)).equals(BigInteger.ONE))
            j = (j.multiply(b)).mod(mod);
        return j;
    }

    public static void main(String[] args) {
        BigInteger e = new BigInteger("10000019");
        BigInteger n = new BigInteger("20000900009");
        BigInteger msg = new BigInteger("20000000000");
        System.out.println("Message: " + msg);
        BigInteger cypher = power(msg, e, n);
        System.out.println("Cypher: " + cypher);

        BigInteger d = new BigInteger("6283524500");
        while (true) {
            BigInteger new_msg = power(cypher, d, n);
            if (new_msg.equals(msg)) {
                System.out.println("New Message: " + new_msg);
                System.out.println("Private key: " + d);
                break;
            }
            d = d.add(BigInteger.ONE);
        }
    }
}

File: Plaintext_Attack.cpp
// Bismillahir Rahmanir Rahim

#include<bits/stdc++.h>
#include<boost/multiprecision/cpp_int.hpp>

using namespace std;
using namespace boost::multiprecision;

cpp_int power(cpp_int b, cpp_int p, cpp_int m) {
    b=b%m;
    if(p==0) return 1;

    cpp_int j=power(b, p/2, m);
    j=(j*j)%m;

    if(p%2==1) j=(j*b)%m;
    return j;
}

cpp_int mod_inv(cpp_int a, cpp_int m) {
    cpp_int org_m=m;
    cpp_int y=0, x=1;

    if(m==1) return 1;

    while(a>1) {
        cpp_int q=a/m;
        cpp_int temp=m;

        m=a%m;
        a=temp;
        temp=y;

        y=x-q*y;
        x=temp;
    }

    if(x<0) x+=org_m;
    return x;
}

int main() {
    cpp_int e=10000019, n=20000900009, m=20000000000;
    cpp_int c=power(m,e,n);
    cpp_int d=6283524500;
    while(1) {
        if(power(c,d,n)==m) break;
        else d++;
    }
    cout<<d<<endl;
}
